R/repoHistory.R

```diff
@@ -0,0 +1,47 @@
+#' Get Git Repository History
+#'
+#' @description
+#' Returns a data frame containing every commit made to the Git repository.
+#' The information recorded includes the commit hash, author, date, and
+#' files changed in each commit.
+#'
+#' @keywords internal
+repoHistory <- function() {
+  # Get the git log with commit hash, author name, author date, and files changed
+  git_log <- processx::run(
+    "git",
+    c(
+      "--literal-pathspecs",          # Treat pathspecs literally
+      "log",
+      "--name-only",                  # Show only names of files changed
+      "--pretty=format:%H\t%an\t%ad", # Customize the log output format
+      "--date=format:%Y-%m-%d %H:%M:%S %z" # Specify date format
+    ),
+    wd = logDir(),
+    echo = FALSE,
+    error_on_status = TRUE
+  )$stdout
```

Comment:
I've taken a look at this `git log` call (at request of @andersone1).  Overall it looks fine to me.  I think the main thing you should think more about is whether it handles paths as you expect when `logDir()` and the top-level of the Git repository don't match (first item below).  The other things are mostly just information and notes; some could be a problem, at least down the line, but are probably fine in this context.

---

 * `git log --name-only ...` outputs paths relative to the *top-level* of the Git repository, regardless of whether it's invoked from a subdirectory.  I'd guess you want to re-root it at `logDir()`.

 * the `--literal-pathspecs` is fine but unnecessary given you're not passing in pathspecs to the `git log` call

 * related to that, perhaps you want to pass some pathspecs?  If callers are going to filter to a specific set of files, it'd be more efficient to let Git do that filtering for you.  You could expose an argument with a good default (say, `c("*.R", "*.ctl", "*.mod", ...others...)` and then tack that onto the `git log` call (`-- {expand here}`).  If you go that route, you need to drop the `--literal-pathspecs` because you want to pass globs.

 * for a branch with a lot of history and/or a lot of files, this slurping and parsing of the whole history could get pretty heavy.  Not something to worry about until you hit it, though.

 * this will also consider renames and deletions as changes to a file (may be what you want; if not, you could consider using `--diff-filter` to narrow down to what you're after)

 * the docs say "every commit made to the Git repository" but this log call is really just for `HEAD`.  I realize that qctools is likely going to be used in a one-branch context, but it still seems like a good idea to update the docs to be clearer (e.g., they can just say "current branch").

 * Git commits have an author date and a committer date.  In the project context (one-branch, likely no cherry picking, amending at most for rewriting, ...), these will usually be identical, and almost always at least be close.  If I were to pick one to use for this function, I'd go with the committer date, I'd go with the committer date as the more important one (but again, it doesn't really matter).

 * I don't spot any issues with the date handling and conversion.

   My preferred way would be to ask Git for the UNIX timestamp (`%at` or `%ct`, depending on whether you want the author or committer time) and then feed that to `as.POSIXct(as.integer(...), origin = "1970-01-01")` (origin is necessary for R < 4.3 compatibility).

   The main appeal of that is 1) it doesn't specify a custom format to `git log` and 2) it involves less conversions on both the Git and R end.  Git stores commit time underneath as the Unix timestamp (in UTC, tz is just for context/display), and R `as.POSIXct` converts things back to a UNIX timestamp.

   To show that this gives the same result, here are the timestamp and your custom format for two Git commits:

   ```
   1724695824 2024-08-26 11:10:24 -0700
   1722324278 2024-07-30 09:24:38 +0200
   ```

   ```R
   as.POSIXct(1724695824)
   #> [1] "2024-08-26 14:10:24 EDT"
   as.POSIXct("2024-08-26 11:10:24 -0700", format = "%Y-%m-%d %H:%M:%S %z")
   #> [1] "2024-08-26 14:10:24 EDT"

   as.POSIXct(1722324278)
   #> [1] "2024-07-30 03:24:38 EDT"
   as.POSIXct("2024-07-30 09:24:38 +0200", format = "%Y-%m-%d %H:%M:%S %z")
   #> [1] "2024-07-30 03:24:38 EDT"
   ```

 * in the project context, you can expect pretty plain and unsurprising file names.  But it's at least worth thinking through what happens with more unusual characters.

   Here's how repoHistory handles a file with spaces (okay, that doesn't count as "more unusual"), a file with a newline, a file with spaces, and a named fileβ.

   ```R
   head(repoHistory()[, "file"], 4)
   #> # A tibble: 4 × 1
   #>   file
   #>   <chr>
   #> 1 "\"file with\\ttab\""
   #> 2 "file with spaces"
   #> 3 "\"file\\316\\262\""
   #> 4 "\"foo\\nwith newline\""
   ```

   That quoting is coming from Git.

   ```
   $ git diff-tree --name-only --no-commit-id HEAD
   "file with\ttab"
   file with spaces
   "file\316\262"
   "foo\nwith newline"
   ```

   It's probably worth tacking on a `-c core.quotePath=false` between `repoHistory`'s `git` and `log` so at least "fileβ" comes back usable (see `man git-config` for more details on `core.quotePath`).

   ```
   $ git -c core.quotePath=false diff-tree --name-only --no-commit-id HEAD
   "file with\ttab"
   file with spaces
   fileβ
   "foo\nwith newline"
   ```

   The most reliable way to handle this when scripting with Git is to use the `-z` flag.  Then Git doesn't do any quoting and separates the names with a NUL byte.  The problem with doing that in the context of R is that I don't know of a way to handle NUL bytes with processx.

   In Python, it'd just be:

   ```python
   >>> import subprocess as sp
   >>> p = sp.run(["git", "diff-tree", "-z", "--no-commit-id", "--name-only", "HEAD"], capture_output = True, encoding = "utf-8")
   >>> p.stdout.split("\0")
   ['file with\ttab', 'file with spaces', 'fileβ', 'foo\nwith newline', '']
   ```

 * this seems to have a lot of overlap with `gitLog` (which takes a list of paths, calls `git log -- $PATH` on each one, and returns combined a data frame with the commits).  Could `gitLog` be replaced by this new function entirely (especially if the function is updated to take a pathspec)?

Reply:
@kyleam Thank you for this thorough review—it’s greatly appreciated. @michaelmcd18 and I will review each item once we return from Thanksgiving break.